### PR TITLE
[WIP] Use seq no for shard recovery

### PR DESCRIPTION
**Pull request by bleskes**
This is a POC PR to gather feedback regarding the approach taken. There are a bunch of nocommits, indicated potential discussion points and things that merit a better solution. If we decide we like the changes I will break this into smaller PRs and add tests etc.
### High level mechanics of seq no based recovery

Using global checkpoints, we have a guarantee of the shard history between a recovery replica and the current primary. The replica removes all operations above the last known global checkpoint and asks the primary to replay all operations it misses. Since we can't roll back to arbitrary points in the history, we rollback to the last lucene commit (later on, NRT if possible) that gives us what we need. This means we will replay operations from before the global check point. If the requested operations are not to be found in the primary translog we fall back to the normal file based sync.

Note that the last point means that _for now_ we can't use seq no based recovery after restart as the primary always flushes after recovering, leaving it's translog empty. We can use sequence numbers to recover a replica that was isolated away from the cluster and rejoined. We can also use them to recover shards that have not been synced-flushed but did go through their global checkpoint sync and have no doc mismatch.
### PR highlights / caveats

1) Add a custom deletion policy that makes sure we keep needed old commits to do rall back.
2) Modify recovery target code to figure out which operations are needed and add the to the recovery requests.
3) Modify recovery source code to track replayed operations and match them to the requested range. If operations are missed, we retry the recovery and explicitly disable seq no based recoveries. Note that long term the plan is to be able to detect this in advance and do the right thing without the need for a retry
4) Persist global checkpoint in the translog. Our original plan was to persist the last known global checkpoint with each lucene commit. This means that when recovering we are likely to roll back a commit point (as we use the last one's checkpoint and roll back to a previous commit). However, persisting the global checkpoint in each translog checkpoint allows us to have better knowledge which makes it much more likely that we will only keep one lucene commit point and be able to use this.
5) Improve recovery testing infra structure to be share more with production code and simulate all the above.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -144,22 +147,36 @@ public InternalEngine(EngineConfig engineConfig) throws EngineException {
...
+                    assert engineConfig.getCommitIdToOpen() == null;
+                    commitToOpen = null;
+                    seqNoStats = new SeqNoStats(NO_OPS_PERFORMED, NO_OPS_PERFORMED, NO_OPS_PERFORMED);
+                } else if (engineConfig.getCommitIdToOpen() != null) {
+                    assert openMode == EngineConfig.OpenMode.OPEN_INDEX_CREATE_TRANSLOG;
+                    commitToOpen = findCommitById(store.directory(), engineConfig.getCommitIdToOpen());
+                    seqNoStats = Store.loadSeqNoStatsFromCommit(commitToOpen);
+                } else {
+                    // nocommit: there should be a better way
+                    commitToOpen = findCommitById(store.directory(), new Store.CommitId(store.readLastCommittedSegmentsInfo().getId()));
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: if you don't specify one IW should open the latest?

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: so if you just open the IW first and get the commit from the writer and laod the stats later it should work?

File: core/src/main/java/org/elasticsearch/index/store/Store.java [hunk 2]
```diff
@@ -267,6 +279,62 @@ public MetadataSnapshot getMetadata(IndexCommit commit) throws IOException {
...
     }
 
+    public IndexCommit findLastCommitBellowSeqNo(final long seqNo) throws IOException {
+        ensureOpen();
+        failIfCorrupted();
+        metadataLock.readLock().lock();
+        IndexCommit lastCommit = null;
+        long commitMaxSeqNo = UNASSIGNED_SEQ_NO;
+        try {
+            for (IndexCommit commit : DirectoryReader.listCommits(directory)) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: can you commetn that `DirectoryReader.listCommits` guarantees commits in ascending order

File: core/src/main/java/org/elasticsearch/index/store/Store.java [hunk 3]
```diff
@@ -267,6 +279,62 @@ public MetadataSnapshot getMetadata(IndexCommit commit) throws IOException {
...
+            if (stats != null) {
+                maxSeqNo = (long) stats.getMaxValue();
+            } else {
+                maxSeqNo = NO_OPS_PERFORMED;
+            }
+        }
+
+        final Map<String, String> commitUserData = commit.getUserData();
+
+        final long localCheckpoint;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: maybe use getOrDefault?

File: core/src/main/java/org/elasticsearch/index/store/Store.java [hunk 4]
```diff
@@ -1398,4 +1466,62 @@ private static long estimateSize(Directory directory) throws IOException {
         }
     }
 
+    public static class CommitId implements Writeable {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: this deserves some javadocs?

File: core/src/main/java/org/elasticsearch/index/store/Store.java [hunk 5]
```diff
@@ -1398,4 +1466,62 @@ private static long estimateSize(Directory directory) throws IOException {
...
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: lots of newlines in here?

File: core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java [hunk 6]
```diff
@@ -33,25 +34,40 @@
 
 /**
  */
-class Checkpoint {
+public class Checkpoint {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java]: make this final? and add javadocs?

File: core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java [hunk 7]
```diff
@@ -33,25 +34,40 @@
...
         this.offset = offset;
         this.numOps = numOps;
         this.generation = generation;
+        this.seqNoGlobalCheckpoint = seqNoGlobalCheckpoint;
     }
 
     Checkpoint(DataInput in) throws IOException {
-        offset = in.readLong();
+        long longOffestAndVersion = in.readLong();
+        byte version = (byte) (longOffestAndVersion >> 56);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java]: why and what ? I am not following?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 8]
```diff
@@ -116,29 +120,57 @@ public RecoverySourceHandler(final IndexShard shard, RecoveryTargetHandler recov
...
+                // at the current indexing edge there may be holes corresponding to operation that are currently
+                // being indexed and those are safe to recovery. Sadly we have no way of knowing which ops these are
+                // instead, we have to verify we have a nice consecutive range.
+                //
+                // nocommit: improving the translog allowing replay all ops from a seqNo up and check it contains a seqNo
+                // will allows us to remove this and also not fail and retry the recovery if we miss to find what we need
+                // but rather immediately start with a file based recovery
+                long currentMax = shard.seqNoStats().getMaxSeqNo();
+                logger.trace("{} waiting for all operations to complete up to [{}]", shard.shardId(), currentMax);
+                cancellableThreads.execute(() -> shard.waitForOpsToComplete(currentMax));
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: this waits for all ops up to _currentMax_ to also be replicated? I couldn't tell from the javadocs?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: no, it waits until all ops up to `currentMax` are finished locally. i.e., the local checkpoint has advanced beyond that point.

File: core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java [hunk 9]
```diff
@@ -33,25 +34,40 @@
...
-            + Long.BYTES;// generation
+            + Long.BYTES // generation
+            + Long.BYTES; // seq no global checkpoint
     final long offset;
     final int numOps;
     final long generation;
+    final long seqNoGlobalCheckpoint;
 
-    Checkpoint(long offset, int numOps, long generation) {
+    Checkpoint(long offset, int numOps, long generation, long seqNoGlobalCheckpoint) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java]: ok so this is just wrong, the `seqNoGlobalCheckpoint` has nothing to do with the checkpoint here. We can't allow arbitrary metadata to be written into this file. This is an 100% no-go. Can we model this differently by maybe adding a special translog operation that gets added into the translog before we sync? so we can just open the translog based on the checkpoint and read the operation from the outside? I don't want to mix concerns here in the translog.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java]: @s1monw and I discussed this and came up with this alternative plan:

1. we have a Translog.setGlobalSeqNoCheckpoint() which updates a field in the current writer
2. we add a special op which contains the current checkpoint and has a fixed known length
3. we write the op before we fsync
4. the offset of the translog after the above write goes in the the checkpoint file
5. we have a static “readGlobalSeqNoCheckpoint” which opens the .ckp file and the corresponding tlog and reads the above struct and returns the value

**Pull request closed without merging**